lib/tasks/manageiq/cypress.rake

```diff
@@ -0,0 +1,69 @@
+namespace :cypress do
+  desc "Run cypress tests"
+  task :ui => 'ui:run'
+
+  namespace :ui do
+    desc "Seed database/configure assets for cypress specs"
+    task :seed do |rake_task|
+      app_prefix = rake_task.name.chomp('cypress:ui:seed')
+      Rake::Task["#{app_prefix}integration:seed"].invoke
+    end
+
+    desc "Interactively run cypress tests"
+    task :open => ['ui:setup'] do
+      sh "#{yarn_cmd} cypress:open"
+    end
+
+    desc "Run headless cypress tests"
+    task :run => ['ui:setup'] do
+      sh "#{yarn_cmd} cypress:run:ci"
+    end
+
+    desc "Stop the backend Rails server"
+    task :stop do |rake_task|
+      app_prefix = rake_task.name.chomp('cypress:ui:stop')
+      Rake::Task["#{app_prefix}integration:stop_server"].invoke
+    end
+
+    task :setup do |rake_task|
+      app_prefix = rake_task.name.chomp('cypress:ui:setup')
+      Rake::Task["#{app_prefix}integration:start_server"].invoke
+    end
+
+    desc "Run cypress tests in 'development' mode"
+    task :dev => 'dev:run'
+
+    # Aliases for running with CYPRESS_DEV
+    namespace :dev do
+      desc "'Development Mode' for cypress:ui:seed"
+      task :seed => [:env, 'cypress:ui:seed']
+
+      desc "'Development Mode' for cypress:ui:open"
+      task :open => [:env, 'cypress:ui:open']
+
+      desc "'Development Mode' for cypress:ui:run"
+      task :run => [:env, 'cypress:ui:run']
+
+      desc "'Development Mode' for cypress:ui:setup"
+      task :setup => [:env, 'cypress:ui:setup']
+
+      desc "'Development Mode' for cypress:ui:stop"
+      task :stop => [:env, 'cypress:ui:stop']
+
+      # Helper task to defaulte the CYPRESS_DEV env var to "1"
+      task :env do
+        ENV["CYPRESS_DEV"] = "1"
+      end
+    end
+
+    def yarn_cmd
+      cmd = "yarn"
+
+      unless defined?(ENGINE_ROOT)
```

Comment:
This is still suspicious, we should run yarn in the UI engine.

If ENGINE_ROOT means different things, why not just use the right engine root?

Reply:
> This is still suspicious, we should run yarn in the UI engine.

This is what is being done, it is just making sure we don't add it if we are running from `manageiq-ui-classic`.

> If `ENGINE_ROOT` means different things, why not just use the right engine root? (`ManageIQ::UI::Classic::Engine.root`)

`ENGINE_ROOT` is defined by the `Rails::Engine` internals, in two spots:

The first is when generating a `Rails::Engine`, and it defines it in the `bin/rails` task:

https://github.com/ManageIQ/manageiq-ui-classic/blob/382d0d19508ac9315b1e5180c977fe16cb6418fe/bin/rails#L5

Secondly, it also defines `ENGINE_ROOT` when the tasks are loaded in the `app:` namespace:

https://github.com/rails/rails/blob/b2eb1d1c55a59fee1e6c4cba7030d8ceb524267c/railties/lib/rails/tasks/engine.rake#L20-L22

This is the one that is defined in these tasks, and only if running from `manageiq-ui-classic` (not `manageiq`).

* * *

However, I agree that we probably ensure we are using `ManageIQ::UI::Classic::Engine.root` for the `--cwd` so we aren't using `ENGINE_ROOT` when we eventually are doing these tests in other providers.  I defaulted to `ENGINE_ROOT` for the `if`, and probably still will, since I don't want to avoid `autoload`ing issue if possible.